codex-rs/execpolicy2/src/parser.rs

```diff
@@ -0,0 +1,238 @@
+use multimap::MultiMap;
+use parking_lot::Mutex;
+use shlex;
+use starlark::any::ProvidesStaticType;
+use starlark::environment::GlobalsBuilder;
+use starlark::environment::Module;
+use starlark::eval::Evaluator;
+use starlark::starlark_module;
+use starlark::syntax::AstModule;
+use starlark::syntax::Dialect;
+use starlark::values::Value;
+use starlark::values::list::ListRef;
+use starlark::values::list::UnpackList;
+use starlark::values::none::NoneType;
+use std::sync::Arc;
+
+use crate::decision::Decision;
+use crate::error::Error;
+use crate::error::Result;
+use crate::policy::validate_match_examples;
+use crate::rule::PatternToken;
+use crate::rule::PrefixPattern;
+use crate::rule::PrefixRule;
+use crate::rule::Rule;
+
+// todo: support parsing multiple policies
+pub struct PolicyParser;
+
+impl PolicyParser {
+    pub fn parse(policy_source: &str, unparsed_policy: &str) -> Result<crate::policy::Policy> {
+        let mut dialect = Dialect::Extended.clone();
+        dialect.enable_f_strings = true;
+        let ast = AstModule::parse(policy_source, unparsed_policy.to_string(), &dialect)
+            .map_err(|e| Error::Starlark(e.to_string()))?;
+        let globals = GlobalsBuilder::standard().with(policy_builtins).build();
+        let module = Module::new();
+
+        let builder = PolicyBuilder::new();
+        {
+            let mut eval = Evaluator::new(&module);
+            eval.extra = Some(&builder);
+            eval.eval_module(ast, &globals)
+                .map_err(|e| Error::Starlark(e.to_string()))?;
+        }
+        Ok(builder.build())
+    }
+}
+
+#[derive(Debug, ProvidesStaticType)]
+struct PolicyBuilder {
+    rules_by_program: Mutex<MultiMap<String, Rule>>,
+}
+
+impl PolicyBuilder {
+    fn new() -> Self {
+        Self {
+            rules_by_program: Mutex::new(MultiMap::new()),
+        }
+    }
+
+    fn add_rule(&self, rule: Rule) {
```

Comment:
For a builder, I don't think we need a `Mutex` because a builder is generally owned by a single thread. Make this `&mut self` instead.